### PR TITLE
[-Wunsafe-buffer-usage] Warning for unsafe invocation of span::data

### PR DESCRIPTION
**Pull request by malavikasamak**
…-Wunsafe-buffer-usage,

there maybe accidental re-introduction of new OutOfBound accesses into the code bases. One such case is invoking span::data() method on a span variable to retrieve a pointer, which is then cast to a larger type and dereferenced. Such dereferences can introduce OutOfBound accesses.

To address this, a new WarningGadget is being introduced to warn against such invocations.

### CODE REVIEW

File: clang/lib/Analysis/UnsafeBufferUsage.cpp [hunk 1]
```diff
@@ -721,6 +721,33 @@ class UnsafeBufferUsageAttrGadget : public WarningGadget {
...
+        Op(Result.Nodes.getNodeAs<ExplicitCastExpr>(OpTag)) {}
+
+  static bool classof(const Gadget *G) {
+    return G->getKind() == Kind::DataInvocation;
+  }
+ 
+  static Matcher matcher() {
+    return stmt(
+        explicitCastExpr(has(cxxMemberCallExpr(callee(
+               cxxMethodDecl(hasName("data")))))).bind(OpTag));
```
Comment (t-rasmud) [inline clang/lib/Analysis/UnsafeBufferUsage.cpp:744]: Will this also match on user defined functions called "data"? I think something like `cxxMethodDecl(ofClass(hasName("span")))` might be needed to match `span.data()` alone (but I might be wrong). In any case, maybe have a test case with a user defined function called "data" which'll show the matcher matches only on `span.data()`.

**Review by t-rasmud — CHANGES_REQUESTED** [review]

File: clang/lib/Analysis/UnsafeBufferUsage.cpp [hunk 2]
```diff
@@ -721,6 +721,33 @@ class UnsafeBufferUsageAttrGadget : public WarningGadget {
   DeclUseList getClaimedVarUseSites() const override { return {}; }
 };
 
+// Warning gadget for unsafe invocation of span::data method.
+// Triggers when the pointer returned by the invocation is immediately
+// cast to a larger type.
+
+class DataInvocationGadget : public WarningGadget {
+  constexpr static const char *const OpTag = "data_invocation_expr";
```
Comment (ziqingluo-90) [inline clang/lib/Analysis/UnsafeBufferUsage.cpp:729]: a nitpick:  I noticed most tag strings use camel cases.   We better be consistent on this?

File: clang/lib/Sema/AnalysisBasedWarnings.cpp [hunk 3]
```diff
@@ -2261,6 +2261,21 @@ class UnsafeBufferUsageReporter : public UnsafeBufferUsageHandler {
         // note_unsafe_buffer_operation doesn't have this mode yet.
         assert(!IsRelatedToDecl && "Not implemented yet!");
         MsgParam = 3;
+      } else if (const auto *ECE = dyn_cast<ExplicitCastExpr>(Operation)) {
+        QualType destType = ECE->getType();
+        const uint64_t dSize = Ctx.getTypeSize(destType.getTypePtr()->getPointeeType());
+        if(const auto *CE =dyn_cast<CXXMemberCallExpr>(ECE->getSubExpr())) {
+
+          if(CE->getRecordDecl()->getQualifiedNameAsString().compare("std::span"))
```
Comment (ziqingluo-90) [inline clang/lib/Sema/AnalysisBasedWarnings.cpp:2269]: I'm a tiny bit skeptical that if the qualified name is always `std::span`.  
Maybe we can have a test like this:
```
using namespace std;
void f(span<int> buf) {
   BigTy *p = (BigTy *) buf.data();
}
```

File: clang/lib/Analysis/UnsafeBufferUsage.cpp (hunk 1 above)
Comment (malavikasamak) [reply clang/lib/Analysis/UnsafeBufferUsage.cpp:744]: Yes. It matches all method invocations with name "data". However, we wouldn't be warning on all of them as the "handleUnsafeOperation" checks if the method belongs to std::span class. It maybe better to move it to the matcher as you point out, so there is less overlap between gadgets.

File: clang/lib/Sema/AnalysisBasedWarnings.cpp (hunk 3 above)
Comment (malavikasamak) [reply clang/lib/Sema/AnalysisBasedWarnings.cpp:2269]: Hmmm. I think it shouldn't be an issue as we are getting the qualified name. However, I will add a test case to check this.

File: clang/lib/Analysis/UnsafeBufferUsage.cpp (hunk 2 above)
Comment (malavikasamak) [reply clang/lib/Analysis/UnsafeBufferUsage.cpp:729]: Thank you. Will fix this.

**Review by t-rasmud — APPROVED** [review]
LGTM!

**Review by ziqingluo-90 — APPROVED** [review]
LGTM!

**Pull request merged**